pallets/eth-bridge/src/lib.rs

```diff
@@ -442,57 +494,89 @@ pub mod pallet {
+                        }
+                    }
+                },
+                Request::Send(_) => {
+                    let tx = req.as_active_tx()?;
+                    let self_is_sender = author.account_id == tx.data.sender;
+                    // Plus 1 for sender
+                    let has_enough_confirmations = util::has_enough_confirmations::<T>(
+                        tx.confirmation.confirmations.len() as u32 + 1u32,
+                    );
+                    if !self_is_sender && !has_enough_confirmations {
+                        let confirmation = eth::sign_msg_hash::<T>(&tx.confirmation.msg_hash)?;
+                        if !tx.confirmation.confirmations.contains(&confirmation) {
+                            call::add_confirmation::<T>(tx.request.tx_id, confirmation, author);
+                        }
+                    } else {
+                        process_active_tx_request::<T>(
                             author,
-                        ),
-                        (None, _) => {},
+                            tx,
+                            self_is_sender,
+                            has_enough_confirmations,
+                        )?;
                     }
+                },
+            }
+        }
+
+        Ok(())
+    }
+
+    fn process_active_tx_request<T: Config>(
+        author: Author<T>,
+        tx: ActiveTransactionData<T>,
+        self_is_sender: bool,
+        tx_has_enough_confirmations: bool,
+    ) -> Result<(), DispatchError> {
+        let tx_is_sent = tx.data.eth_tx_hash != H256::zero();
+        let tx_is_past_expiry = util::time_now::<T>() > tx.data.expiry;
+
+        if self_is_sender && tx_has_enough_confirmations && !tx_is_sent {
+            let lock_name = format!("eth_bridge_ocw::send::{}", tx.request.tx_id).as_bytes().to_vec();
+            let mut lock = AVN::<T>::get_ocw_locker(&lock_name);
+
+            // Protect against sending more than once
+            if let Ok(guard) = lock.try_lock() {
+                let eth_tx_hash = eth::send_tx::<T>(&tx)?;
+                call::add_eth_tx_hash::<T>(tx.request.tx_id, eth_tx_hash, author);
+                guard.forget(); // keep the lock so we don't send again
+            } else {
+                log::info!(
+                    "👷 Skipping sending txId: {:?} because ocw is locked already.",
+                    tx.request.tx_id
+                );
+            };
+        } else if !self_is_sender && (tx_is_sent || tx_is_past_expiry) {
+            if util::requires_corroboration::<T>(&tx.data, &author) {
+                match eth::corroborate::<T>(&tx, &author)? {
+                    (Some(status), tx_hash_is_valid) => call::add_corroboration::<T>(
+                        tx.request.tx_id,
+                        status,
+                        tx_hash_is_valid.unwrap_or_default(),
+                        author,
+                    ),
+                    (None, _) => {},
```

Comment:
So the expiry is for the corroborating aspect as well? That is, if we fail to get validation from the other collators, even while the transaction has been sent and successfully executed on Ethereum, T2 will think it has failed?

Reply:
Ok, I think we may have an edge case here. 
**Assumptions**: the expiry period (timeout) is the total time T2 waits for confirmation of a T1 transaction (my initial assumption was that it is the last time ETHEREUM will accept the Tx as valid, which is slightly different)

**Case 1:**: Tx Failed on Ethereum
This is a benign case. Because we only store positive corroborations, no one will vote for this Tx so T2 will eventually mark it as failed due to the timeout.

**Case 2**: Tx successful on Ethereum with normal slack
This is a benign case. Because of our BFT assumptions, there will always be at least 2/3 collators who are not Byzantine. This  means not only that they say the truth (they're not malicious) but also that they are up to date. That is, being Byzantine also includes networking failures and not outright lying or deceitful communication. If 2/3 of the nodes are not Byzantine, they will actively try to corroborate and send an answer as soon as they can.

**Case 2**: Tx successful on Ethereum but very late
This is an edge case that may be dangerous. This means the Tx went in before expiry was passed, _but it could have been very close to the edge_. This is where the problem comes in: even though we have 2/3 or more of honest collators, they simply may not have enough time to do their work: notice the tx, go ask for corroboration and put the result back in the chain, as that is done via OCWs and takes a couple of blocks at least. That is around 30 seconds, I guess.

In this case, there will be no votes in the chain for this Tx, and if we were close to the Expiry limit when it went in, T2 may well reject it due to timeout for lack of votes, even though the TX was processed.

Reply:
Discussed offline: some of the assumptions I had are not true. I'm happy with the merge.